sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardRecordsIterator.java

```diff
@@ -77,20 +77,28 @@ public ShardRecordsIterator(final ShardCheckpoint initialCheckpoint,
 
   private void readMoreIfNecessary() throws TransientKinesisException {
     if (data.isEmpty()) {
-      GetKinesisRecordsResult response;
+      GetKinesisRecordsResult response = fetchRecords();
+      data.addAll(filter.apply(response.getRecords(), checkpoint));
+    }
+  }
+
+  private GetKinesisRecordsResult fetchRecords() throws TransientKinesisException {
+    GetKinesisRecordsResult response = null;
+    do {
       try {
         response = kinesis.getRecords(shardIterator, checkpoint.getStreamName(),
             checkpoint.getShardId());
+        shardIterator = response.getNextShardIterator();
```

Comment:
Why do you get new shard iterator each time?

Reply:
ShardIterator has to be advanced after successful response in order to iterate over a shard. Isn't that correct?

Reply:
Sorry, my bad. Please ignore my comment.